server/src/pages/shot/editor-history.js

```diff
@@ -0,0 +1,113 @@
+const { Selection } = require("../../../shared/selection");
+
+exports.EditorHistory = class {
+  constructor(devicePixelRatio) {
+    this.beforeEdits = [];
+    this.afterEdits = [];
+    this.devicePixelRatio = devicePixelRatio;
+  }
+
+  push(canvas, area, recordType) {
+    const record = new EditRecord(
+      canvas,
+      area,
+      this.devicePixelRatio,
+      recordType
+    );
+    this.beforeEdits.push(record);
+    this.afterEdits = [];
+  }
+
+  pushDiff(canvas, area) {
+    this.push(canvas, area, RecordType.DIFF);
+  }
+
+  pushFrame(canvas, area) {
+    this.push(canvas, area, RecordType.FRAME);
+  }
+
+  canUndo() {
+    return !!this.beforeEdits.length;
+  }
+
+  undo(canvasBeforeUndo) {
```

Comment:
Both undo and redo takes in the canvas before the action is applied because we need to save the image or part of it to revert the action (undo a redo or redo an undo).